src/ApiException.php

```diff
@@ -78,6 +78,84 @@ public function getStatus()
         return $this->status;
     }
 
+    public function getDomain()
+    {
+        $metadata = $this->metadata;
+        // For the createFromRpcStatus() method, $metadata is an object and not an array so 'containsErrorInfo' causes a type error.
+        if (is_object($metadata)) {
+            $metadataRpcStatus = Serializer::decodeAnyMessages($metadata);
+            $detailsRpcStatus = self::containsErrorInfo($metadataRpcStatus);
+            if ($detailsRpcStatus['containsErrorInfo']) {
+                return $metadataRpcStatus[$detailsRpcStatus['keyOfErrorInfo']]['domain'];
+            }
+        } elseif (is_Array($metadata)) {
+            // For REST-based responses, the $metadata does not need to be decoded.
+            $detailsRest = self::containsErrorInfo($metadata);
+            if ($detailsRest['containsErrorInfo']) {
+                return $metadata[$detailsRest['keyOfErrorInfo']]['domain'];
+            }
+        }
```

Comment:
This `if (is_object) elseif (is_Array)` logic as well as the decoding logic is in multiple spots, that means it is a good thing to try to either 1. pull into a function 2. process ahead of time so that these getters just need to access a variable.

Can we try to refactor this code into one of the two approaches?

Reply:
Good point. I've refactored into a separate function.